counsel.el

```diff
@@ -5222,37 +5222,37 @@ subdirectories that builds may be invoked in."
            counsel-compile-build-directories))
 
 (defun counsel--get-build-subdirs (blddir)
-  "Return all subdirs of BLDDIR sorted by access time."
+  "Return all subdirs of BLDDIR sorted by modification time."
   (mapcar #'car (sort (directory-files-and-attributes
                        blddir t directory-files-no-dot-files-regexp t)
                       (lambda (x y)
-                        ;; FIXME: Access time is NOT the 7th file attribute.
                         (time-less-p (nth 6 y) (nth 6 x))))))
 
 (defun counsel-compile-get-build-directories (&optional dir)
   "Return a list of potential build directories."
   (let* ((srcdir (or dir (counsel--compile-root)))
-         (blddir (counsel--find-build-subdir srcdir))
-         (props `(srcdir ,srcdir blddir ,blddir recursive t)))
+         (blddir (counsel--find-build-subdir srcdir)))
     (mapcar (lambda (s)
-              (setq s (concat (propertize "Select build in "
-                                          'face 'font-lock-warning-face)
-                              (propertize s 'face 'dired-directory)))
-              (add-text-properties 0 (length s) props s)
-              s)
+              (propertize
+               (concat (propertize "Select build in "
+                                   'face 'font-lock-warning-face)
+                       (propertize s 'face 'dired-directory))
+               'blddir s 'srcdir srcdir 'recursive t))
             (and blddir (counsel--get-build-subdirs blddir)))))
 
-;; No easy way to make directory local, would buffer local make more sense?
+;; This is a workaround for the fact there is no concept of "project"
+;; local variables (as opposed to for example buffer-local). So we
+;; store all our history in a global list filter out the results we
+;; don't want.
 (defun counsel-compile-get-filtered-history (&optional dir)
   "Return a compile history relevant to current project."
   (let ((root (or dir (counsel--compile-root)))
         history)
     (dolist (item counsel-compile-history)
       (let ((srcdir (get-text-property 0 'srcdir item))
             (blddir (get-text-property 0 'blddir item)))
-        ;; FIXME: File names are not regexps!
-        (when (or (and srcdir (string-match-p srcdir root))
-                  (and blddir (string-match-p blddir root)))
+        (when (or (and srcdir (file-in-directory-p srcdir root))
+                  (and blddir (file-in-directory-p blddir root)))
```

Comment:
Shouldn't the arguments to `file-in-directory-p` be flipped?

Reply:
I don't think so... 

> Signature
(file-in-directory-p FILE DIR)

Documentation
Return non-nil if FILE is in DIR or a subdirectory of DIR.

Reply:
Here's why I ask:
```el
(string-match-p "/home/blc/.emacs.d/"      "/home/blc/.emacs.d/")      ; => 0
(string-match-p "/home/blc/.emacs.d/elpa/" "/home/blc/.emacs.d/")      ; => nil
(string-match-p "/home/blc/.emacs.d/"      "/home/blc/.emacs.d/elpa/") ; => 0

(file-in-directory-p "/home/blc/.emacs.d/"      "/home/blc/.emacs.d/")      ; => t
(file-in-directory-p "/home/blc/.emacs.d/elpa/" "/home/blc/.emacs.d/")      ; => t
(file-in-directory-p "/home/blc/.emacs.d/"      "/home/blc/.emacs.d/elpa/") ; => nil
```

As you can see, the sense of the arguments to `string-match-p` is the reverse of that for `file-in-directory-p`. Perhaps the `string-match-p` implementation was wrong to begin with; I'm not familiar with the relation of `srcdir`, `blddir`, and `root`.

Reply:
It looks like we just get away with it in the common case but there is a bug when blddir was a subdir which gets elided by the fact srcdir happened to be the same. The sense of the file-in-directory-p tests is correct all the time.